GitUI/UserControls/RevisionGrid/Graph/RevisionGraph.cs

```diff
@@ -308,59 +371,56 @@ public void Insert(GitRevision workTreeRev, GitRevision indexRev, IReadOnlyList<
-        /// is not in the same index in the orderednodecache, the order has been changed. Only then rebuilding is
-        /// required. If the order is changed after this revision, we do not care since it wasn't processed yet.
-        /// But when filtering revisions, this can mismatch. So check the first revision in addition.
+        /// Check if the row cache is dirty by comparing to the ordered node cache.
         /// </summary>
-        private static bool IsRowCacheDirty(IList<RevisionGraphRow> orderedRowCache, RevisionGraphRevision[] orderedNodesCache)
+        /// <param name="orderedNodesCache">The ordered nodes cache.</param>
+        /// <param name="orderedRowCache">The ordered row cache.</param>
+        /// <returns><see langword="true"/> if the row cache is dirty; otherwise <see langword="false"/>.</returns>
+        private bool IsRowCacheDirty(ImmutableArray<RevisionGraphRow> orderedRowCache, ImmutableArray<RevisionGraphRevision> orderedNodesCache)
         {
-            // We need bounds checking on orderedNodesCache. It should be always larger then the rowcache,
+            // We need bounds checking on orderedNodesCache. It should be always larger then the orderedRowCache,
             // but another thread could clear the orderedNodesCache while another is building orderedRowCache.
             // This is not a problem, since all methods use local instances of those caches. We do need to invalidate.
-            if (orderedRowCache.Count > orderedNodesCache.Length)
+            if (_rowCacheInvalid || orderedRowCache.Length > orderedNodesCache.Length)
             {
                 return true;
             }
 
-            if (orderedRowCache.Count == 0)
+            if (orderedRowCache.Length == 0)
             {
                 return false;
             }
 
-            int indexToCompare = orderedRowCache.Count - 1;
+            // It is very easy to check if the orderedRowCache is dirty or not. If the last revision added to the orderedRowCache
+            // is not in the same index in the orderedNodesCache, the order has been changed. Only then rebuilding is
+            // required. If the order is changed after this revision, we do not care since it wasn't processed yet.
+            // But when filtering revisions, this can mismatch. So check the first revision in addition.
+            int indexToCompare = orderedRowCache.Length - 1;
             return orderedRowCache[indexToCompare].Revision != orderedNodesCache[indexToCompare]
                 || orderedRowCache[0].Revision != orderedNodesCache[0];
         }
 
-        private void BuildOrderedRowCache(RevisionGraphRevision[] orderedNodesCache, int currentRowIndex, int lastToCacheRowIndex)
+        private void BuildOrderedRowCache(int currentRowIndex, int lastToCacheRowIndex)
         {
+            ImmutableArray<RevisionGraphRevision> orderedNodesCache = BuildOrderedNodesCache(currentRowIndex);
```

Comment:
`currentRowIndex` is only used here now. Rather have `ImmutableArray<RevisionGraphRevision> orderedNodesCache` as input argument.